### ISSUE TITLE
Using the memory and SQLite backends in a multi-threaded/multiprocess app on PythonAnywhere

### ISSUE THREAD

**Issue by Avasam** [reclosedev/requests-cache#337]
My Python app is hosted on PythonAnywhere.
I have a maximum of 1GB of storage (around 800MBs left for SQLite) and up to 3GB of available RAM (before my process gets rebooted).
I have tried memorizing my requests in the past, but while it worked fine locally, I found it didn't seem to work in production (I now understand it's because of the different webworkers spawned). Plus it'd eventually just memory leak.
So far SQLite has been working flawlessly, but I found it tends to slow down quite a lot as the cache size increases. It's also much slower than the in-memory implementation I had (it's not disk throttling as my NVMe SSD runs at 6-12% during that time).

In memory has some clear advantages: max size, speed, and it gets automatically cleared with a reboot and an email notification if things get out of hands, and my cache can be volatile (I already use `fast_save`). But my current understanding is that it will not work for my needs (keep the cache across multiple users using my app in parallel, in the span of 5-20 minutes). If so, is there a better (faster) solution than SQLite I could use?

Source code: https://github.com/Avasam/speedrun.com_global_scoreboard_webapp/blob/a2bb8ed7c3d1440719f386f2308643ce0436a7fb/services/utils.py

**Comment by JWCook** [comment]
That's an interesting use case! For in-memory, multi-threaded, multi-process access, Redis would generally be your best option, although I don't know if that's possible with PythonAnywhere. `memcached` would be another good option; requests-cache doesn't support that right now, but It's been on my list of features to look into for awhile.

The memory 'backend' is really just a dict, and more of a placeholder than anything else. It could be made thread-safe by throwing a `threading.RLock` on top of it, but multiprocess would be a lot more work. I'd like to investigate other options first, but if nothing else works for you, we could look into making a multiprocess-compatible in-memory cache.

**Comment by JWCook** [comment]
I'm surprised about SQLite slowing down for you; it should be able to handle terabytes of data without breaking a sweat. I'd like to dig into that some more, if you're interested. Could you run some benchmarks for me to see what kind of speeds you're getting? Example:
```python
from time import perf_counter as time
from requests_cache import CachedSession

ITERATIONS = 10000
CACHE_ITEM_SIZE = 1024 * 6

def test_write_speed(cache):
    start = time()
    for i in range(ITERATIONS):
        cache[f'key_{i}'] = b'1' * CACHE_ITEM_SIZE
    elapsed = time() - start
    avg = (elapsed / ITERATIONS) * 1000
    print(f'Elapsed: {elapsed:.3f} (avg {avg:.3f}ms per write)')

session = CachedSession(backend='sqlite')
test_write_speed(session.cache.responses)
```

You could run that once with an empty cache, and once with a 500MB+ cache. You can use this script to generate one: [generate_test_db.py](https://github.com/reclosedev/requests-cache/blob/master/examples/generate_test_db.py)

SQLite also supports in-memory databases, which could be an option if disk space is limited.

**Comment by JWCook** [comment]
Updated those scripts to work with the latest release. Locally I get about 2-3ms per write (tested on an empty cache and a 4GB cache). Let me know what you get in a PythonAnywhere environment.

**Comment by Avasam** [comment]
I don't think redis is an option on PythonAnywhere either. I was looking at memcached as well before I found requests-cache (the implementation was pretty much plug&play, I spent more time cleaning up old code that became obsolete)

My bad I didn't specify, my issue is actually with cached read speeds (especially since I'm using `fast_save`). I'll still run your test. And I'll run a read test. Just to make sure to isolate it in case I *somehow* did something that affected cached requests speed.

Here are some example DB files: https://www.dropbox.com/s/vgjvoxyhb5zhmri/http_cache.7z?dl=0  

One of the request should be:
https://www.speedrun.com/api/v1/runs?user=qjn1wzw8&status=verified&embed=level%2Cgame.levels%2Cgame.variables&max=200
I restored file 1-3 from my Recycle Bin, but I believe both ~700MB files did not exclude any parameters from the key.

**Comment by Avasam** [comment]
Write tests results:
```ps
PS > py .\examples\benchmark.py
Elapsed: 18.794 (avg 3.759ms per write)
PS > py .\examples\generate_test_db.py
Generating responses... ---------------------------------------- 100% 0:00:00
(took about 1m)
PS > py .\examples\benchmark.py
Elapsed: 21.313 (avg 4.263ms per write)
```

I was gonna [write up a nice small post](https://pastebin.com/LWVe9tqy) about the differences between my real world performance and the test. And as I was trying to remove more and more possible variables that could be caused by the code it clicked:
Some veeeeery old hacky check I put in because of PythonAnywhere's thread count limitations (of about 8 per worker/process).
Turns out unlimited threads is 13x faster than 8 huh, 🤔 Who would've known /s

Anyway, problem explained. Probably another artefact of my college-level coding skills at the time. (Or truly a limitation of my host I'll have to come up with a creative solution)

If in-memory cache is a possibility, I would still gladly take it. I don't really care about it being process-safe (race conditions) as long as it works (no deadlock).  But the importance of this suggestion for me definitely went way down.
I don't plan to keep the cache for more than a day (maybe up to a week as a pipedream, or if I shell-out for more storage).  
Splitting the cache into that "long lived shared" and "short lived per user as a safety for interruptions" you mentioned in #21 will probably also help

**Comment by JWCook** [comment]
Glad that worked out! Sure, go ahead and open up a new issue as a feature request. That just helps make these issues more easily searchable by other users.

> Splitting the cache into that "long lived shared" and "short lived per user as a safety for interruptions" you mentioned in #21 will surely also help

I'm not seeing any mention of that in #21. Is that from a different issue?

**Comment by Avasam** [comment]
No idea why I tagged that issue, I meant this comment: https://github.com/reclosedev/requests-cache/issues/336#issuecomment-890359067
